COMPILER.md

```diff
@@ -0,0 +1,286 @@
+            .set_frontend_opt(dialect > 21);
+        let unopt_res = compile_file(
+            &mut allocator,
+            runner.clone(),
+            opts.clone(),
+            &input_program,
+            &mut symbol_table,
+        );
+        let res = if do_optimize {
+            unopt_res.and_then(|x| run_optimizer(&mut allocator, runner, Rc::new(x)))
+        } else {
+            unopt_res.map(Rc::new)
+        };
+
+This the the full lifecycle of chialisp compilation from an outside perspective.
+
+If you want to parse the source file yourself, you can use parse_sexp from 
+src/compiler/sexp, which yields a result of Vec&lt;Rc&lt;SExp&gt;&gt;, a vector
+of refcounted pointers to s-expression objects.  These are richer than clvm
+values in that atoms, strings, hex values,integers and nils are distinguishable
+in what's read.  This is similar to the code in src/classic/clvm/type.rs but
+since these value distinctions are first-class in SExp, all values processed
+by the compiler, starting with the preprocessor (src/compiler/preprocessor.rs), 
+going through the frontend (src/compile/frontend.rs) and to code generation
+(src/compiler/codegen.rs) use these values.  Every expression read from the
+input stream has a Srcloc (src/compiler/srcloc.rs) which indicates where it
+was read from in the input, and as long as the srcloc is copied from form to
+form, it allows the compiler to maintain known associations between input and
+results.  For example, you can implement syntax coloring by simply using
+parse\_sexp (src/compiler/sexp.rs) to parse a file, run it through the
+preprocessor and frontend, and then decorate the locations present in the
+HelperForms accessible from the CompilerForm that results.  If an error is
+returned, it contains a srcloc that's relevant.
+
+You can break down the basics of how chialisp compilation functions like this:
+
+        let mut allocator = Allocator::new();
+        let mut symbol_table = HashMap::new();
+        let runner = Rc::new(DefaultProgramRunner::new());
+        let opts = Rc::new(DefaultCompilerOpts::new(&use_filename))
+            .set_optimize(do_optimize)
+            .set_search_paths(&search_paths)
+            .set_frontend_opt(dialect > 21);
+        let parsed = parse_sexp(Srcloc::start(&use_filename), file_content.bytes())?;
+        let compileform = frontend(opts.clone(), &parsed)?;
+        let generated_code = codegen(&mut allocator, runner, opts, &compileform, &mut symbol_table)?;
+        // generated_code.1 is an Rc<SExp> which contains the output
+        // code.
+
+Everything before code generation is uninteresting, but I'll note at a high level
+how functions on PrimaryCodegen function.
+
+codegen starts by running start_codegen to introduce each helper to the
+PrimaryCodegen and based on their types, bin them into the appropriate parts
+of the code generator to lookup later:
+
+    Defconstants are turned into mods and run, unlike classic chialisp.
```

Comment:
```suggestion
    Defconstants are turned into mods and run.
```
Only software archeologists care about classic.

Reply:
I included some comparisons since I do get questions along these lines (especially in PRs about things that make changes to cl21).